src/server/cluster/cluster_config.cc

```diff
@@ -25,29 +27,58 @@ std::string_view ClusterConfig::KeyTag(std::string_view key) {
   return key.substr(start + 1, end - start - 1);
 }
 
-SlotId ClusterConfig::KeySlot(std::string_view key) {
-  std::string_view tag = KeyTag(key);
+SlotId ClusterConfig::KeySlot(string_view key) {
+  string_view tag = KeyTag(key);
   return crc16(tag.data(), tag.length()) & kMaxSlotNum;
 }
 
-ClusterConfig::ClusterConfig() {
+ClusterConfig::ClusterConfig(string_view my_id) : my_id_(my_id) {
   cluster_enabled = true;
-  AddSlots();
 }
 
-void ClusterConfig::AddSlots() {
-  // TODO update logic acording to config
-  // currently add all slots to owned slots
-  std::lock_guard lk{slots_mu_};
-  for (SlotId slot_id = 0; slot_id <= kMaxSlotNum; ++slot_id) {
-    owned_slots_.emplace(slot_id);
+namespace {
+optional<ClusterConfig::Node> GetMasterNodeForSlot(const vector<ClusterConfig::Node>& nodes,
+                                                   const string& my_id) {
+  for (const auto& node : nodes) {
+    if (node.id == my_id) {
+      // We do not store information for our own node, as it's unneeded, and it also helps us to
+      // determine our ownership quicker.
+      return nullopt;
+    }
+
+    if (node.role == ClusterConfig::Role::kMaster) {
+      return node;
+    }
+  }
+
+  CHECK(false) << "No master present in nodes";
+  return nullopt;
+}
+}  // namespace
+
+void ClusterConfig::SetConfig(const vector<ClusterShard>& new_config) {
+  lock_guard gu(slots_mu_);
+
+  for (const auto& shard : new_config) {
+    optional<Node> node = GetMasterNodeForSlot(shard.nodes, my_id_);
+
+    for (const auto& slot_range : shard.slot_ranges) {
+      for (SlotId i = slot_range.start; i <= slot_range.end; ++i) {
+        config_[i] = node;
+      }
+    }
   }
-  return;
 }
 
-bool ClusterConfig::IsMySlot(SlotId id) {
-  std::shared_lock sl(slots_mu_);
-  return owned_slots_.contains(id);
+bool ClusterConfig::IsMySlot(SlotId id) const {
+  return GetNodeForSlot(id) == nullopt;
```

Comment:
We will need to validate that all slots are allocated in the given config, otherwise this logic will fail

Reply:
Currently (pre-this PR), when the process starts it assumes that it owns all slots. This logic maintains that, but I actually think it should be the other way around, i.e. that the process should start with 0 slots and take on new slots as needed.
This is worth discussing actually, but anyway, as long as config requests are full, we are good. We can enforce that in `SetConfig()`, like you proposed above.

Reply:
yes you are right, when the server starts it will start with 0 slots, this is also what I wrote in the design doc. I just created this logic in my PR so that I could check my logic in all unit test with cluster_mode yes as default. But we can fix this now